### PR TITLE
SALTO-1153: include-list-based-fetch

### PR DESCRIPTION
**Pull request by alonstern**
Changed the format of the fetch configuration in salesforce to include based according to the MVP suggestion in https://www.notion.so/saltoio/Include-List-Based-Fetches-41bfef27e64e43a893afc5405a6c1c3e.

Created configuration suggestion for a user using the old format to upgrade the new format.

---
_Release Notes_: 
Salesforce Adapter:
The configuration options `metadataTypesSkippedList`, `instancesRegexSkippedList` and `dataManagement` were replaced with a new option `fetch` to allow more flexibility in filtering the data to fetch. On the next fetch, Salto will automatically upgrade the old options to the new option (with the user approval).

### CODE REVIEW

File: packages/salesforce-adapter/src/adapter_creator.ts [hunk 1]
```diff
@@ -48,71 +48,90 @@ const credentialsFromConfig = (config: Readonly<InstanceElement>): Credentials =
...
-  validateClientConfig(config?.value?.client)
-  const adapterConfig: { [K in keyof Required<SalesforceConfig>]: SalesforceConfig[K] } = {
-    metadataTypesSkippedList: makeArray(config?.value?.metadataTypesSkippedList),
-    instancesRegexSkippedList: makeArray(config?.value?.instancesRegexSkippedList),
-    maxItemsInRetrieveRequest: config?.value?.maxItemsInRetrieveRequest,
-    dataManagement: config?.value?.dataManagement,
-    useOldProfiles: config?.value?.useOldProfiles,
-    client: config?.value?.client,
+  try {
+    validateDeprecatedParameters(config)
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/adapter_creator.ts:102]: wouldn't it be easier and cleaner to update the config structure before so the validation won't have to take the old format into account? it will be dead code basically on every run except for the first run in a couple of existing workspaces.

that way it will also be easier to remove the unused code once we can remove support for the old configuration format (there would be no need to change the validation code, unlike with this version where the validation code is explicitly aware of the two formats)

File: packages/salesforce-adapter/src/config_change.ts [hunk 2]
```diff
@@ -27,6 +27,9 @@ const MESSAGE_INTRO = 'Salto failed to fetch some items from salesforce. '
 const MESSAGE_REASONS_INTRO = 'Due to the following issues: '
 const MESSAGE_SUMMARY = 'In order to complete the fetch operation, '
 + 'Salto needs to stop managing these items by applying the following configuration change:'
+export const DEPRECATED_OPTIONS_MESSAGE = 'The configuration options "metadataTypesSkippedList", "instancesRegexSkippedList" and "dataManagement" are deprecated.'
++ ' The following changes will update the deprected options to the "fetch" configuration option.'
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/config_change.ts:31]:
```suggestion
+ ' The following changes will update the deprecated options to the "fetch" configuration option.'
```

might want to install a [spell check plugin](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) to catch these kinds of typos more easily

File: packages/salesforce-adapter/src/config_change.ts [hunk 3]
```diff
@@ -86,24 +89,104 @@ export const createRetrieveConfigChange = (result: RetrieveResult): ConfigChange
...
+const convertDeprecatedRegex = (filePathRegex: string): string => {
+  let newPathRegex = filePathRegex
+
+  newPathRegex = filePathRegex.startsWith('^')
+    ? newPathRegex.substring(1)
+    : newPathRegex
+
+  newPathRegex = !filePathRegex.startsWith('.*') && !filePathRegex.startsWith('^')
+    ? `.*${newPathRegex}`
+    : newPathRegex
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/config_change.ts:101]: checking `filePathRegex` but then using `newPathRegex` as a value seems like it might be a bug?
`newPathRegex` might have a different value than `filePathRegex` at this point.
seems strange to check one value and then infer what it means on another value (depends on all previous operations, which makes it very hard to follow and verify)

File: packages/salesforce-adapter/src/config_change.ts [hunk 4]
```diff
@@ -86,24 +89,104 @@ export const createRetrieveConfigChange = (result: RetrieveResult): ConfigChange
...
+        }
+        return { metadataType: convertDeprecatedRegex(`${regexParts[0]}$`), name: convertDeprecatedRegex(`^${regexParts.slice(1).join('.')}`) }
+      }),
+    ...makeArray(deprecatedParams.metadataTypesSkippedList).map(type => ({ metadataType: type })),
+  ]
+
+  const includes = makeArray(deprecatedParams.instancesRegexSkippedList)
+    .includes(PACKAGES_INSTANCES_REGEX)
+    ? [{ namespace: '', name: '.*', metadataType: '.*' }]
+    : []
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/config_change.ts:152]: not sure I understand this.
IIUC this is saying that if the current configuration says we should skip all packages except for the standard, then we will add the standard package to the `include` config, but if we don't have that filter we will not include the standard package?
does that mean that we won't fetch any of the standard package metadata?
do we even need such an option to not fetch the standard package?

File: packages/salesforce-adapter/src/config_change.ts [hunk 5]
```diff
@@ -86,24 +89,104 @@ export const createRetrieveConfigChange = (result: RetrieveResult): ConfigChange
...
+export const convertDeprecatedDataConf = (conf: DataManagementConfig): DataManagementConfig => {
+  const updatedConf = _.cloneDeep(conf)
+  updatedConf.includeObjects = updatedConf.includeObjects.map(convertDeprecatedRegex)
+  updatedConf.excludeObjects = updatedConf?.excludeObjects?.map(convertDeprecatedRegex)
+  updatedConf.allowReferenceTo = updatedConf?.allowReferenceTo?.map(convertDeprecatedRegex)
+  updatedConf.saltoIDSettings.overrides = updatedConf.saltoIDSettings.overrides?.map(
+    override => ({ ...override, objectsRegex: convertDeprecatedRegex(override.objectsRegex) })
+  )
+  return updatedConf
+}
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/config_change.ts:123]: less side effects and mutation in-place, avoid `cloneDeep` whenever possible
```suggestion
export const convertDeprecatedDataConf = (conf: DataManagementConfig): DataManagementConfig => ({
  ...conf,
  includeObjects: conf.includeObjects.map(convertDeprecatedRegex),
  excludeObjects: conf?.excludeObjects?.map(convertDeprecatedRegex),
  allowReferenceTo: conf?.allowReferenceTo?.map(convertDeprecatedRegex)
  saltoIDSettings: {
    ...conf.saltoIDSettings,
    overrides: conf.saltoIDSettings.overrides?.map(
      override => ({ ...override, objectsRegex: convertDeprecatedRegex(override.objectsRegex) })
    ),
  },
})
```

File: packages/salesforce-adapter/src/config_change.ts [hunk 6]
```diff
@@ -86,24 +89,104 @@ export const createRetrieveConfigChange = (result: RetrieveResult): ConfigChange
...
-  const currentDataManagement = currentConfig.dataManagement
-  const metadataTypesSkippedList = makeArray(configChangesByType.metadataTypesSkippedList)
-    .map(e => e.value)
-    .filter(e => !currentMetadataTypesSkippedList.includes(e))
-  const instancesRegexSkippedList = makeArray(configChangesByType.instancesRegexSkippedList)
+): { config: InstanceElement ; message: string } | undefined => {
+  const currentMetadataExclude = makeArray(currentConfig.fetch?.metadata?.exclude)
+
+  const currentDataManagement = currentConfig.fetch?.data
+    ?? (currentConfig.dataManagement && convertDeprecatedDataConf(currentConfig.dataManagement))
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/config_change.ts:173]: wouldn't it be better to convert the current config to the new format in one place and allow this code to assume the configuration is already in the correct format?
again, the benefit would be that the code here can be written such that it won't need to change when we remove the code that handles the format upgrade and it will be much easier to follow and review because it won't have to constantly handle the format upgrade as well as everything it is actually supposed to do

File: packages/salesforce-adapter/src/custom_object_instances_deploy.ts [hunk 7]
```diff
@@ -306,7 +306,7 @@ const isModificationChangeList = <T>(
 export const deployCustomObjectInstancesGroup = async (
   changes: ReadonlyArray<Change<InstanceElement>>,
   client: SalesforceClient,
-  dataManagementConfig?: DataManagementConfig,
+  dataManagement?: DataManagement,
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/custom_object_instances_deploy.ts:309]: why remove the `config` part for the name?
just calling it `dataManagement` makes it sound like it is an object that is responsible for implementing or interfacing with data management operations when in fact it is still just a configuration

File: packages/salesforce-adapter/src/fetch.ts [hunk 8]
```diff
@@ -159,19 +169,23 @@ type RetrieveMetadataInstancesArgs = {
...
   maxItemsInRetrieveRequest,
-  instancesRegexSkippedList,
+  metadataQuery,
 }: RetrieveMetadataInstancesArgs): Promise<FetchElements<InstanceElement[]>> => {
   const configChanges: ConfigChangeSuggestion[] = []
 
   const notInSkipList = (file: FileProperties): boolean => (
-    !instancesRegexSkippedList.some(re => re.test(`${file.type}.${file.fullName}`))
+    metadataQuery.isInstanceMatch({
+      namespace: file.namespacePrefix === undefined || file.namespacePrefix === '' ? DEFAULT_NAMESPACE : file.namespacePrefix,
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/fetch.ts:185]: extract to function? `getNamespace` that gets file properties and returns the namespace?

File: packages/salesforce-adapter/e2e_test/workflow.test.ts [hunk 9]
```diff
@@ -182,7 +183,7 @@ describe('workflow filter', () => {
         baseCustomObject)
       expect(rawWorkflowInstance).toBeDefined()
       fetchResult = [...rawWorkflowTypes, rawWorkflowInstance as InstanceElement]
-      await runFiltersOnFetch(client, {}, fetchResult)
+      await runFiltersOnFetch(client, { fetchProfile: buildFetchProfile({}) }, fetchResult)
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/salesforce-adapter/e2e_test/workflow.test.ts:186]: `defaultFilterContext` is already used insided `runFiltersOnFetch` so I changed `runFiltersOnFetch` to accept `Partial<FilterContext>` so I won't need to pass `buildFetchProfile({})` here

File: packages/salesforce-adapter/src/config_change.ts (hunk 3 above)
Comment (alonstern) [reply packages/salesforce-adapter/src/config_change.ts:101]: I want to convert `^abc` to `abc` and `bcd` to `.*bcd`. If in each one of these checks, I will check `newPathRegex`. `^abc` will be converted to `.*abc`, so I always need to check the original regex.

File: packages/salesforce-adapter/src/config_change.ts (hunk 4 above)
Comment (alonstern) [reply packages/salesforce-adapter/src/config_change.ts:152]: If we don't have that filter, we won't change anything in the include list. I.e., if `metadata.includes` wasn't defined before in the configuration (which means to include everything) we will leave it undefined (because of the `pickBy` in line 163) and if it was defined we won't change the list.

File: packages/salesforce-adapter/src/custom_object_instances_deploy.ts (hunk 7 above)
Comment (alonstern) [reply packages/salesforce-adapter/src/custom_object_instances_deploy.ts:309]: There are `DataManagementConfig` which really contains the configuration and `DataManagement` which contains the operations.

File: packages/salesforce-adapter/src/adapter_creator.ts (hunk 1 above)
Comment (alonstern) [reply packages/salesforce-adapter/src/adapter_creator.ts:102]: The main problem with that is that the messages of errors of values in old fields will be shown on the fields

File: packages/salesforce-adapter/src/config_change.ts (hunk 6 above)
Comment (alonstern) [reply packages/salesforce-adapter/src/config_change.ts:173]: This function must get the `currentConfig` not already updated because it needs to know whether an update is needed or not, so I must make changes in this function. I did reorganize the code a little so all the code related to deprecated fields will be next to each other

File: packages/salesforce-adapter/src/adapter_creator.ts [hunk 10]
```diff
@@ -48,71 +48,90 @@ const credentialsFromConfig = (config: Readonly<InstanceElement>): Credentials =
...
+
+    try {
+      validateFetchParameters(config?.value?.[FETCH_CONFIG] ?? {})
+    } catch (e) {
+      if (e instanceof ConfigValidationError) {
+        e.fieldPath.unshift(FETCH_CONFIG)
+      }
+      throw e
```
[the remark this answers is not in the document]
Comment (alonstern) [reply packages/salesforce-adapter/src/adapter_creator.ts:110]: I need the reason for an error and the path to the field that caused the error. Just to make sure, are you suggesting that instead of throwing an exception each validation function will return a message and path if there was an error or undefined if there wasn't and I will check after each validation if the results are undefined? (Note that it will cause many more if conditions since in some places I wrap multiple validations under one try-catch)

File: packages/salesforce-adapter/src/deprecated_config.ts [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+// Based on the list in https://salesforce.stackexchange.com/questions/101844/what-are-the-object-and-field-name-suffixes-that-salesforce-uses-such-as-c-an
+const INSTANCE_SUFFIXES = [
+  'c', 'r', 'ka', 'kav', 'Feed', 'ViewStat', 'VoteStat', 'DataCategorySelection', 'x', 'xo', 'mdt', 'Share', 'Tag',
+  'History', 'pc', 'pr', 'hd', 'hqr', 'hst', 'b', 'latitude__s', 'longitude__s', 'e', 'p', 'ChangeEvent', 'chn',
+]
+export const PACKAGES_INSTANCES_REGEX = `^.+\\.(?!standard_)[^_]+__(?!(${INSTANCE_SUFFIXES.join('|')})([^a-zA-Z\\d_]+|$)).+$`
+
+const DEPRECATED_FIELDS = [
+  INSTANCES_REGEX_SKIPPED_LIST,
```
Comment (ori-moisis) [inline packages/salesforce-adapter/src/deprecated_config.ts:41]: should these constants be removed from the `types.ts` file and moved here?

**Review by ori-moisis — APPROVED** [review]

File: packages/salesforce-adapter/src/deprecated_config.ts (hunk 11 above)
Comment (alonstern) [reply packages/salesforce-adapter/src/deprecated_config.ts:41]: It creates a dependency cycle: `deprecated_config.ts` needs some constants and types from `types.ts` (e.g., `FetchParameters`) and `types.ts` need these constants for the `configType` definition
